enclave-tls/samples/sgx-stub-enclave/sgx_stub_ecall.c

```diff
@@ -0,0 +1,283 @@
+	if (sgx_status != SGX_SUCCESS || sockfd < 0) {
+		ETLS_ERR("Failed to call socket() %#x %d\n",
+                         sgx_status, sockfd);
+		return -1;
+	}
+
+	struct etls_sockaddr_in s_addr;
+	memset(&s_addr, 0, sizeof(s_addr));
+	s_addr.sin_family = ETLS_AF_INET;
+	s_addr.sin_addr.s_addr = s_ip;
+	s_addr.sin_port = s_port;
+
+	/* Connect to the server */
+	int ocall_ret = 0;
+	sgx_status = ocall_connect(&ocall_ret,
+                                   sockfd,
+                                   (struct etls_sockaddr *)&s_addr,
+                                   sizeof(s_addr));
+	if (sgx_status != SGX_SUCCESS || ocall_ret == -1) {
+		ETLS_ERR("failed to call connect() %#x %d\n",
+                          sgx_status, ocall_ret);
+		return -1;
+	}
+
+	/* Enclave-tls init */
+	libenclave_tls_init();
+	enclave_tls_handle handle;
+	enclave_tls_err_t ret = enclave_tls_init(&conf, &handle);
+	if (ret != ENCLAVE_TLS_ERR_NONE) {
+		ETLS_ERR("Failed to initialize enclave tls %#x\n", ret);
+		return -1;
+	}
+
+	ret = enclave_tls_negotiate(handle, (int)sockfd);
+	if (ret != ENCLAVE_TLS_ERR_NONE) {
+		ETLS_ERR("Failed to negotiate %#x\n", ret);
+		goto err;
+	}
+
+	const char *msg = "Hello and welcome to Enclave TLS!\n";
+	size_t len = strlen(msg);
+	ret = enclave_tls_transmit(handle, (void *)msg, &len);
+	if (ret != ENCLAVE_TLS_ERR_NONE || len != strlen(msg)) {
+		ETLS_ERR("Failed to transmit %#x\n", ret);
+		goto err;
+	}
+
+	char buf[256];
+	len = sizeof(buf);
+	ret = enclave_tls_receive(handle, buf, &len);
+	if (ret != ENCLAVE_TLS_ERR_NONE) {
+		ETLS_ERR("Failed to receive %#x\n", ret);
+		goto err;
+	}
+
+	if (len >= sizeof(buf))
+		len = sizeof(buf) - 1;
+	buf[len] = '\0';
+
+	/* Server running in SGX Enclave will send mrenclave, mrsigner and hello message to client */
+	if (len >= 2 * sizeof(sgx_measurement_t)) {
+		ETLS_INFO("Server's SGX identity:\n");
+		ETLS_INFO("  . MRENCLAVE = ");
+		for (int i = 0; i < 32; ++i)
+			printf("%02x", (uint8_t)buf[i]);
+		printf("\n");
+		ETLS_INFO("  . MRSIGNER  = ");
+		for (int i = 32; i < 64; ++i)
+			printf("%02x", (uint8_t)buf[i]);
+		printf("\n");
+
+		memcpy(buf, buf + 2 * sizeof(sgx_measurement_t),
+                       len - 2 * sizeof(sgx_measurement_t));
+		buf[len - 2 * sizeof(sgx_measurement_t)] = '\0';
+
+		ETLS_INFO("Server:\n%s\n", buf);
+	} else {
+	        /* Server not running in SGX Enlcave will only send hello message to client */
+	        ETLS_INFO("Server: %s\n", buf);
+	}
+
+	/* Sanity check whether the response is expected */
+	if (strcmp(msg, buf)) {
+	        printf("Invalid response retrieved from enclave-tls server\n");
+	        goto err;
+	}
+
+	ret = enclave_tls_cleanup(handle);
+	if (ret != ENCLAVE_TLS_ERR_NONE)
+	        ETLS_ERR("Failed to cleanup %#x\n", ret);
+
+	return ret;
+
+err:
+        /* Ignore the error code of cleanup in order to return the prepositional error */
+        enclave_tls_cleanup(handle);
+        return -1;
+}
```

Comment:
Is it possible to reuse the logic of ecall_etls_client_startup and ecall_etls_server_startup for both SGX and non-SGX build?

Refer to how QVL and QVE in DCAP share the common implementation of sgx_qve_verify_quote():
- The shared codes between them are implemented in SGXDataCenterAttestationPrimitives/QuoteVerification/Qve/Enclave/qve.cpp.
- QVL: use conditional macro for `#define sgx_qve_verify_quote sgx_qvl_verify_quote` and declare sgx_qvl_verify_quote() in its header file.
- QVE: directly use sgx_qve_verify_quote()
- As what you already did in etls_enclave_attester_load_all.c, use wrapper functions to call e.g, socket() and ocall_socket(), for non-SGX and SGX builds.

Reply:
I saw this, but it seems that all ocall function in ecall will be implemented like this